pkg/quarterdeck/db/models/users.go

```diff
@@ -1035,6 +1035,74 @@ func (u *User) fetchRoles(tx *sql.Tx) (err error) {
 	return rows.Err()
 }
 
+const (
+	getUserOrgRoleSQL = "SELECT r.name FROM organization_users ur JOIN roles r ON ur.role_id=r.id WHERE user_id=:userID AND organization_id=:orgID"
+	getNumOwnersSQL   = "SELECT COUNT(*) FROM organization_users WHERE organization_id=:orgID and role_id IN (SELECT id from roles where name=:ownerRole)"
+	userRoleUpdateSQL = "UPDATE organization_users SET modified=:modified, role_id=(SELECT id FROM roles WHERE name=:role) WHERE user_id=:userID AND organization_id=:orgID"
+)
+
+// ChangeRole updates the role of the user in specified organization.
+func (u *User) ChangeRole(ctx context.Context, orgID any, role string) (err error) {
+	// Validate the orgID
+	var userOrg ulid.ULID
+	if userOrg, err = ulids.Parse(orgID); err != nil {
+		return err
+	}
+
+	if ulids.IsZero(userOrg) {
+		return ErrMissingOrgID
+	}
+
+	// Validate the role
+	if !perms.IsRole(role) {
+		return ErrInvalidRole
+	}
+
+	var tx *sql.Tx
+	if tx, err = db.BeginTx(ctx, nil); err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	u.SetModified(time.Now())
+
+	// Get the user's current role
+	var currentRole string
+	if err = tx.QueryRow(getUserOrgRoleSQL, sql.Named("userID", u.ID), sql.Named("orgID", userOrg)).Scan(&currentRole); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return ErrUserOrganization
+		}
+		return err
+	}
+
+	// Make sure that the organization has at least one owner if the user is being
+	// changed to a non-owner role
+	if currentRole == perms.RoleOwner && role != perms.RoleOwner {
+		var numOwners int
+		if err = tx.QueryRow(getNumOwnersSQL, sql.Named("orgID", userOrg), sql.Named("ownerRole", perms.RoleOwner)).Scan(&numOwners); err != nil {
+			return err
+		}
+
+		switch numOwners {
+		case 0:
+			return ErrNoOwnerRole
+		case 1:
+			return ErrOwnerRoleConstraint
+		}
```

Comment:
Case 0 shouldn't happen really - if it does, the database has been corrupted; so hopefully we'll see `ErrNoOwnerRole` in Sentry and take immediate action.

Reply:
And the logic does make sense; this is saying that numOwners has to be 2 or greater (or negative I suppose … but that won't be returned from the query).